man/man3/fopenoutput.3
.TH fopenoutput 3  "LPJmL programmers manual"
.SH NAME
fopenoutput \- creates LPJmL output files
.SH SYNOPSIS
.nf
\fB#include "lpj.h"

Outputfile *fopenoutput(const Cell \fIgrid[]\fB,int \fInout\fB,const Config *\fIconfig\fB);\fP

.fi
.SH DESCRIPTION
Function creates and opens output files specified in the LPJmL configuration.
The arguments of \fBfopenoutput\fP are:
.TP
.I grid
Pointer to LPJmL grid. Must be initialized by \fBnewgrid()\fP.
.TP
.I config
Pointer to LPJmL configuration data. Must be read by \fBreadconfig()\fP.
.TP
.I nout
Maximum number of output files.
.SH RETURN VALUE
Upon successful completion pointer to allocated output data is returned or NULL in case of failure.
.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
fcloseoutput(3), readconfig(3), newgrid(3), iterate(3), isopen(3) 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

## [5.6.25] - 2023-07-29

### Added

- New outputs added for N fluxes from timber extraction and for N pools in wood product pools
- New outputs added for `estab_storage` C and N
- Added missing N fluxes, `estab_storage`, and reservoirs to `globalflux` output
- Added new fluxes to man pages

### Fixed 

- Corrected/added units in comments in output.h
- Corrected units in `flux_sum` man page
- Corrected `N2O_DENIT` output, which now no longer also includes `N2O_NIT` values
- Corrected writing of mineral N pools to outputs, so that also mineral N in frozen soil layers is included
- Corrected writing of `VEGN` for crops, so that `bm_inc.nitrogen` is no longer included to avoid double accounting

### Removed

- Removed unecessary initialization of output variables to zero in `update_annual.c`, which is done in `initoutputdata.c`


## [5.6.24] - 2023-07-29

### Fixed

-  The number of bands is now correctly checked for the `read_residue_data` input.


## [5.6.23] - 2023-07-28

### Changed

- Instead of the soil code the index of the grid cell is written into the NetCDF file for `"grid"` output.


## [5.6.22] - 2023-07-27

### Changed

- Configuration files `lpjml_netcdf.js`, `lpjml_non.js`, `param_non.js`, and `lpjparam_non.js` updated to latest lpjml version.

### Removed

- Predefined daily outputs (`"d_lai"`-`"d_pet"`) and corresponding functions removed.

- Version numbers removed from man pages.

### Fixed

- Variable `issocket` always initialized to `FALSE` in `readfilename.c`.


## [5.6.21] - 2023-07-21

### Changed

- Flag `"nitrogen_coupled"` disabled for lpjml offline runs. Water stress is always coupled with nitrogen stress. Only if lpjml is compiled with `-DCOUPLING_WITH_FMS` flag is enabled.

- Flag `"new_trf"` renamed to `"transp_suction_fcn"`.


## [5.6.20] - 2023-07-13

### Added

- New flags `"fix_co2"` and `"fix_co2_year"` added.

### Changed

- Flag `"istimber"` renamed to `"luc_timber"`.

- Flag `"new_phenology"` renamed to `"gsi_phenology"`.

- Flag `"shuffle_climate"` renamed to `"shuffle_spinup_climate"`.

- The settings for fixed input data have been updated. For climate and N deposition input an interval can be specified from which data is taken after the specified year. In can specified whether random shuffling or cycling is used:

```java
"fix_climate" : false,                /* fix climate after specified year */
"fix_climate_year" : 1901,            /* year after climate is fixed */
"fix_climate_interval" : [1901,1930],
"fix_climate_shuffle" : true,          /* randomly shuffle climate in the interval */
"fix_deposition_with_climate" : false, /* fix N deposition same as climate */
"fix_deposition" : false,              /* fix N deposition after specified year */
"fix_deposition_year" : 1901,          /* year after deposition is fixed */
"fix_deposition_interval" : [1901,1930],
"fix_deposition_shuffle" : true,       /* randomly shuffle depositions in the interval */
"fix_landuse" : false,                 /* fix land use after specfied year */
"fix_landuse_year" : 1901,             /* year after land use is fixed */
"fix_co2" : false,                     /* fix atmospheric CO2  after specfied year */
"fix_co2_year" : 1901,                 /* year after CO2 is fixed */
```

### Removed

- Obsolete flag `"const_climate"` removed.


## [5.6.19] - 2023-07-06

### Added

- Help option `-h` added to `bin2cdf` and `clm2cdf` utility.

- Target `lpjcheck` added to `Makefile` to create only the `lpjcheck` utility.

### Changed

- The version of all man pages has been set to the LPJmL version number, date of last modification removed.

- The different versions of the utilities have been replaced by a reference to the LPJmL version.

### Removed

- All html files have been removed from the repository.

- Obsolete link `lpj` to `lpjml` removed.


## [5.6.18] - 2023-07-04

### Fixed

- Soil array `wpwp` is now correctly initialized with `soilpar->wpwp` for the `"prescribed_soilpar"` setting.


## [5.6.17] - 2023-06-30

### Fixed

- Replaced logical `or` by logical `and` in `survive.c`. This solves issue [#309](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/309).


## [5.6.16] - 2023-06-30

### Fixed

- Removed excess `pft->nind` in all harvest routines in `harvest_stand.c`.


## [5.6.15] - 2023-06-30

### Added

- Check for matching size of discharge queues read from restart file added to `initriver()` function in `initdrain.c`. If the size differs from the value calculated from the river lengths the queues will be resized and initialized to zero. Error message is additionally printed.


## [5.6.14] - 2023-06-08

### Removed

- Support for the deprecated `*.conf` configuration file format has been removed. Only JSON files are allowed for configuration files and metafiles for input.

- Man pages for the description of the `*.conf` file format removed.

### Changed

- Strings in JSON configuration files can now be longer than 256 characters.


## [5.6.13] - 2023-05-16

### Added

- Missing update of `"harvestc"` and `"harvestn"` output added in `annual_biomass_tree.c`, `annual_woodplantation.c`, `daily_agriculture_tree.c`, and `annual_agriculture_tree.c`.


## [5.6.12] - 2023-05-12

### Added


src/lpj/fopenoutput.c
/**************************************************************************************/
/**                                                                                \n**/
/**                  f  o  p  e  n  o  u  t  p  u  t  .  c                         \n**/
/**                                                                                \n**/
/**     C implementation of LPJmL                                                  \n**/
/**                                                                                \n**/
/**     Function creates files for output                                          \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

static Bool create(Netcdf *cdf,const char *filename,int index,
                   Coord_array *array,const Config *config)
{
  int size;
  size=outputsize(config->outputvars[index].id,
                  config->npft[GRASS]+config->npft[TREE],
                  config->npft[CROP],config);
  if(size==1)
    return create_netcdf(cdf,filename,
                         (config->outputvars[index].id==GRID) ? "cellid" :
                         config->outnames[config->outputvars[index].id].var,
                         (config->outputvars[index].id==GRID) ? "cell id" :
                         config->outnames[config->outputvars[index].id].descr,
                         (config->outputvars[index].id==GRID) ? "" :
                         config->outnames[config->outputvars[index].id].unit,
                         (config->outputvars[index].id==GRID) ? LPJ_INT : getoutputtype(config->outputvars[index].id,FALSE),
                         getnyear(config->outnames,config->outputvars[index].id),
                         (config->outnames[config->outputvars[index].id].timestep==ANNUAL) ? 1 : config->outnames[config->outputvars[index].id].timestep,array,config);
  else
    return create_pft_netcdf(cdf,filename,
                             config->outputvars[index].id,
                             config->npft[GRASS]+config->npft[TREE],
                             config->npft[CROP],
                             config->outnames[config->outputvars[index].id].var,
                             config->outnames[config->outputvars[index].id].descr,
                             config->outnames[config->outputvars[index].id].unit,
                             getoutputtype(config->outputvars[index].id,FALSE),
                             getnyear(config->outnames,config->outputvars[index].id),
                             (config->outnames[config->outputvars[index].id].timestep==ANNUAL) ? 1 : config->outnames[config->outputvars[index].id].timestep,array,config);
} /* of 'create' */

static void openfile(Outputfile *output,const Cell grid[],
                     const char *filename,int i,
                     const Config *config)
{
  Header header;
  if(config->outputvars[i].filename.fmt==CDF)
  {
    switch(config->outputvars[i].id)
    {
      case ADISCHARGE: case GRID:
        if(output->index_all==NULL)
        {
          output->index_all=createcoord_all(grid,config);
          if(output->index_all==NULL)
          {
            output->files[config->outputvars[i].id].isopen=FALSE;
            return;
          }
        }
        break;
      default:
        if(output->index==NULL)
        {
          output->index=createcoord(output,grid,config);
          if(output->index==NULL)
          {
            output->files[config->outputvars[i].id].isopen=FALSE;
            return;
          }
        }
        break;
    } /* of switch */
    if(isroot(*config))
    {
      if(!config->outputvars[i].oneyear)
      {
        if(!config->ischeckpoint && config->outputvars[i].filename.meta)
          fprintoutputjson(i,0,config);
        if(!create(&output->files[config->outputvars[i].id].fp.cdf,filename,i,
                   (config->outputvars[i].id==ADISCHARGE || config->outputvars[i].id==GRID) ? output->index_all : output->index,config))
          output->files[config->outputvars[i].id].isopen=TRUE;
      }
    }
  }
  else if(isroot(*config) && !config->outputvars[i].oneyear)
  {
    if(!config->ischeckpoint && config->outputvars[i].filename.meta)
      fprintoutputjson(i,0,config);
    switch(config->outputvars[i].filename.fmt)
    {
       case CLM:
        if(config->ischeckpoint && config->outputvars[i].id!=GRID  && config->outputvars[i].id!=COUNTRY && config->outputvars[i].id!=REGION)
        {
          if((output->files[config->outputvars[i].id].fp.file=fopen(filename,"r+b"))==NULL)
            printfopenerr(config->outputvars[i].filename.name);
          else
          {
            output->files[config->outputvars[i].id].isopen=TRUE;
            if(config->checkpointyear>=config->outputyear)
            {
              fseek(output->files[config->outputvars[i].id].fp.file,
                    headersize(LPJOUTPUT_HEADER,config->outputvars[i].filename.version)+
                    getsize(i,config)*(config->checkpointyear-config->outputyear+1),SEEK_SET);
            }
            else
              fseek(output->files[config->outputvars[i].id].fp.file,
                    headersize(LPJOUTPUT_HEADER,config->outputvars[i].filename.version),SEEK_SET);
          }
        }
        else
        {
          if((output->files[config->outputvars[i].id].fp.file=fopen(filename,"wb"))==NULL)
            printfcreateerr(config->outputvars[i].filename.name);
          else
          {
            output->files[config->outputvars[i].id].isopen=TRUE;
            header.firstyear=config->outputyear;
            if(config->outputvars[i].id==ADISCHARGE)
              header.ncell=config->nall;
            else
              header.ncell=config->total;
            header.firstcell=config->firstgrid;
            header.cellsize_lon=(float)config->resolution.lon;
            header.cellsize_lat=(float)config->resolution.lat;
            header.scalar=1;
            if(config->outputvars[i].id==GRID)
            {
              if(config->float_grid)
                header.datatype=LPJ_FLOAT;
              else
              {
                header.datatype=LPJ_SHORT;
                header.scalar=0.01;
              }
              header.nbands=2;
              header.nstep=1;
              header.timestep=1;
              header.nyear=1;
              header.order=CELLYEAR;
              fwriteheader(output->files[config->outputvars[i].id].fp.file,
                           &header,LPJGRID_HEADER,config->outputvars[i].filename.version);
            }
            else
            {
              header.order=CELLSEQ;
              if(config->outputvars[i].id==COUNTRY || config->outputvars[i].id==REGION)
              {
                header.nstep=1;
                header.timestep=1;
                header.nyear=1;
              }
              else
              {
                header.nstep=getnyear(config->outnames,config->outputvars[i].id);
                header.timestep=max(1,config->outnames[config->outputvars[i].id].timestep);
                header.nyear=(config->lastyear-config->outputyear+1)/header.timestep;
                header.firstyear=config->outputyear+header.timestep-1;
              }
              header.nbands=outputsize(config->outputvars[i].id,
                                       config->npft[GRASS]+config->npft[TREE],
                                       config->npft[CROP],config);
              header.datatype=getoutputtype(config->outputvars[i].id,FALSE);
              fwriteheader(output->files[config->outputvars[i].id].fp.file,
                           &header,LPJOUTPUT_HEADER,config->outputvars[i].filename.version);
            }
          }
        }
        break;
      case RAW:
        if(config->ischeckpoint && config->outputvars[i].id!=GRID  && config->outputvars[i].id!=COUNTRY && config->outputvars[i].id!=REGION)
        {
          if((output->files[config->outputvars[i].id].fp.file=fopen(filename,"r+b"))==NULL)
            printfopenerr(config->outputvars[i].filename.name);
          else
          {
            output->files[config->outputvars[i].id].isopen=TRUE;
            if(config->checkpointyear>=config->outputyear)
            {
              fseek(output->files[config->outputvars[i].id].fp.file,
                    getsize(i,config)*(config->checkpointyear-config->outputyear+1),SEEK_SET);
            }
          }
        }
        else
        {
          if((output->files[config->outputvars[i].id].fp.file=fopen(filename,"wb"))==NULL)
            printfcreateerr(config->outputvars[i].filename.name);
          else
            output->files[config->outputvars[i].id].isopen=TRUE;
        }
        break;
      case TXT:
        if((output->files[config->outputvars[i].id].fp.file=fopen(filename,(config->ischeckpoint && config->outputvars[i].id!=GRID  && config->outputvars[i].id!=COUNTRY && config->outputvars[i].id!=REGION) ? "a" : "w"))==NULL)
          printfcreateerr(config->outputvars[i].filename.name);
        else
          output->files[config->outputvars[i].id].isopen=TRUE;
        break;
    } /* of switch */
  }
  output->files[config->outputvars[i].id].oneyear=config->outputvars[i].oneyear;
} /* of 'openfile' */

Outputfile *fopenoutput(const Cell grid[],   /**< LPJ grid */
                        int n,               /**< size of output file array */
                        const Config *config /**< LPJmL configuration */
                       )                     /** \return output file data or NULL */
{
  int i,size;
#ifdef USE_MPI
  int count;
#endif
  int ncell;
  char *filename;
  Outputfile *output;
  output=new(Outputfile);
  check(output);
  output->files=newvec(File,n);
  check(output->files);
  output->n=n;
  output->index=output->index_all=NULL; 
  for(i=0;i<n;i++)
    output->files[i].isopen=output->files[i].issocket=FALSE;
#ifdef USE_MPI
  output->counts=newvec(int,config->ntask);
  check(output->counts);
  output->offsets=newvec(int,config->ntask);
  check(output->offsets);
  count=config->count;
  MPI_Allgather(&count,1,MPI_INT,output->counts,1,MPI_INT,
                config->comm);
  /* calculate array offsets */
  output->offsets[0]=0;
  for(i=1;i<config->ntask;i++)
    output->offsets[i]=output->offsets[i-1]+output->counts[i-1];
#endif
  outputnames(output,config);
  for(i=0;i<config->n_out;i++)
  {
    if(hassuffix(config->outputvars[i].filename.name,config->compress_suffix))
    { 
      output->files[config->outputvars[i].id].compress=TRUE;
      filename=stripsuffix(config->outputvars[i].filename.name);
      check(filename);
    }
    else
    {
      output->files[config->outputvars[i].id].compress=FALSE;
      filename=config->outputvars[i].filename.name;
    }
    output->files[config->outputvars[i].id].filename=config->outputvars[i].filename.name;
    output->files[config->outputvars[i].id].fmt=config->outputvars[i].filename.fmt;
    if(iscoupled(*config) && config->outputvars[i].filename.issocket)
    {
      output->files[config->outputvars[i].id].issocket=TRUE;
      output->files[config->outputvars[i].id].id=config->outputvars[i].filename.id;
      if(isroot(*config))
      {
        if(config->outputvars[i].id==GLOBALFLUX)
          ncell=0;
        else
          ncell=(config->outputvars[i].id==ADISCHARGE) ? config->nall : config->total;
        if(config->outputvars[i].id==GLOBALFLUX)
          size=sizeof(Flux)/sizeof(Real);
        else if(config->outputvars[i].id==GRID)
          size=2;
        else
          size=outputsize(config->outputvars[i].id,
                          config->npft[GRASS]+config->npft[TREE],
                          config->npft[CROP],config);
        if(openoutput_coupler(config->outputvars[i].filename.id,ncell,getnyear(config->outnames,config->outputvars[i].id),size,getoutputtype(config->outputvars[i].id,config->float_grid),config))
        {
          output->files[config->outputvars[i].id].issocket=FALSE;
          fprintf(stderr,"ERROR100: Cannot open socket stream for output '%s'.\n",
                  config->outnames[config->outputvars[i].id].name);
        }
      }
#ifdef USE_MPI
      MPI_Bcast(&output->files[config->outputvars[i].id].issocket,1,MPI_INT,
                0,config->comm);
#endif
    }
    if(config->outputvars[i].filename.fmt!=SOCK)
      openfile(output,grid,filename,i,config);
#ifdef USE_MPI
    MPI_Bcast(&output->files[config->outputvars[i].id].isopen,1,MPI_INT,
              0,config->comm);
#endif
    if(output->files[config->outputvars[i].id].compress)
      free(filename);
  }
  return output;
} /* of 'fopenoutput' */

void openoutput_yearly(Outputfile *output,int year,const Config *config)
{
  char *filename;
  Header header;
  int i,size,count;
  for(i=0;i<config->n_out;i++)
    if(config->outputvars[i].oneyear)
    {
      if(isroot(*config))
      {
        count=snprintf(NULL,0,config->outputvars[i].filename.name,year);
        if(count==-1)
        {
          fprintf(stderr,"ERROR247: Invalid format in filename '%s'.\n",config->outputvars[i].filename.name);
          output->files[config->outputvars[i].id].isopen=FALSE;
        }
        else
        {
          filename=malloc(count+1);
          check(filename);
          snprintf(filename,count+1,config->outputvars[i].filename.name,year);
          if(config->outputvars[i].filename.meta)
            fprintoutputjson(i,year,config);
          switch(config->outputvars[i].filename.fmt)
          {
            case CLM:
              if((output->files[config->outputvars[i].id].fp.file=fopen(filename,"wb"))==NULL)
              {
                printfcreateerr(filename);
                output->files[config->outputvars[i].id].isopen=FALSE;
              }
              else
              {
                output->files[config->outputvars[i].id].isopen=TRUE;
                header.firstyear=year;
                header.nyear=1;
                header.ncell=config->total;
                header.firstcell=config->firstgrid;
                header.cellsize_lon=(float)config->resolution.lon;
                header.cellsize_lat=(float)config->resolution.lat;
                header.scalar=1;
                header.order=CELLSEQ;
                header.timestep=1;
                header.nstep=getnyear(config->outnames,config->outputvars[i].id);
                header.nbands=outputsize(config->outputvars[i].id,
                                         config->npft[GRASS]+config->npft[TREE],
                                         config->npft[CROP],config);
                if(config->outputvars[i].id==SDATE || config->outputvars[i].id==HDATE || config->outputvars[i].id==SEASONALITY)
                  header.datatype=LPJ_SHORT;
                else
                  header.datatype=LPJ_FLOAT;
                fwriteheader(output->files[config->outputvars[i].id].fp.file,
                             &header,LPJOUTPUT_HEADER,config->outputvars[i].filename.version);

              }
              break;
          case RAW:
            if((output->files[config->outputvars[i].id].fp.file=fopen(filename,"wb"))==NULL)
            {
              printfcreateerr(filename);
              output->files[config->outputvars[i].id].isopen=FALSE;
            }
            else
              output->files[config->outputvars[i].id].isopen=TRUE;
            break;
          case TXT:
            if((output->files[config->outputvars[i].id].fp.file=fopen(filename,"w"))==NULL)
            {
              output->files[config->outputvars[i].id].isopen=FALSE;
              printfcreateerr(filename);
            }
            else
              output->files[config->outputvars[i].id].isopen=TRUE;
            break;
          case CDF:
            size=outputsize(config->outputvars[i].id,
                            config->npft[GRASS]+config->npft[TREE],
                            config->npft[CROP],config);
           if(size==1)
             output->files[config->outputvars[i].id].isopen=!create1_netcdf(&output->files[config->outputvars[i].id].fp.cdf,filename,
                           config->outnames[config->outputvars[i].id].var,
                           config->outnames[config->outputvars[i].id].descr,
                           config->outnames[config->outputvars[i].id].unit,
                           getoutputtype(config->outputvars[i].id,FALSE),
                           getnyear(config->outnames,config->outputvars[i].id),
                           (config->outputvars[i].id==ADISCHARGE) ? output->index_all : output->index,year,config);
           else
             output->files[config->outputvars[i].id].isopen=!create1_pft_netcdf(&output->files[config->outputvars[i].id].fp.cdf,filename,
                           config->outputvars[i].id,
                           config->npft[GRASS]+config->npft[TREE],
                           config->npft[CROP],
                           config->outnames[config->outputvars[i].id].var,
                           config->outnames[config->outputvars[i].id].descr,
                           config->outnames[config->outputvars[i].id].unit,
                           getoutputtype(config->outputvars[i].id,FALSE),
                           getnyear(config->outnames,config->outputvars[i].id),year,output->index,config);

          } /* of switch */
          free(filename);
        }
     } /* of(isroot(*config)) */
#ifdef USE_MPI
     MPI_Bcast(&output->files[config->outputvars[i].id].isopen,1,MPI_INT,0,config->comm);
#endif
   }
} /* of 'openoutput_yearly */
